### ISSUE TITLE
Policy based on GrantType/Client

### ISSUE THREAD

**Issue by StefanKoenigMUC** [IdentityServer/IdentityServer4#4144]
Hi all,

so far all my approaches with IdentityServer went quite well, until today I encountered a problem with policies I was unable to resolve by myself.

I've to clients configured

```
    return new List<Client>
        {
            new Client
            {
                [...]
                AllowedGrantTypes = GrantTypes.Implicit,
                [....]
            },
            new Client
            {
                [...]
                AllowedGrantTypes = GrantTypes.ClientCredentials,
                [....]
            }
        };
```

and a controller like this
```
   [Route("api/forms")]
    [ApiController]
    [Authorize(Policy = "user.api.portfolio.manager")]
    [Authorize(Policy = "application.api.portfolio.manager")]
   public class FormsController : ControllerBase
   {
      [...]
   }
```

and policies like this:

 ```
      private System.Action<AuthorizationOptions> AddJwtAuthorizationPolicyForRole()
        {
            return options => { options.AddPolicy("**POLICY_FOR_GRANT_IMPLICIT**", policy => { 
                policy.AuthenticationSchemes.Add(JwtBearerDefaults.AuthenticationScheme); 
                policy.RequireAuthenticatedUser();
                policy.RequireClaim(ClaimTypes.Role, "USER_ACCESSIBLE");
            }); 
            };
        }

        private System.Action<AuthorizationOptions> AddJwtAuthorizationPolicyForRole()
        {
            return options => { options.AddPolicy("**POLICY_FOR_CLIENT_CREDENTIALS**", policy => { 
                policy.AuthenticationSchemes.Add(JwtBearerDefaults.AuthenticationScheme); 
                policy.RequireAuthenticatedUser();
            }); 
            };
        }
```

but with this configuration, neither of those clients can authenticate as both policies must be true.

Is there any example how my case can be resolved (i think this should be a quite common pattern,  but was unable to figure out how to resolve it..)

Basically: 
2 Clients
1 Controller which should be accessed
Accessibility is different for each of those client types.

Thanks in advance for your help,

**Comment by fuzzzerd** [comment]
If you need to do "OR" logic within a Policy, you'll need to dig into [ASP.NET Policy Authorization](https://docs.microsoft.com/en-us/aspnet/core/security/authorization/policies?view=aspnetcore-3.1#why-would-i-want-multiple-handlers-for-a-requirement).

**Comment by StefanKoenigMUC** [comment]
tried that, but then I've still the issue that I need 2 policies and have to fiddle around with requirement interactions ("if policy == that ignore that rule, else apply that rule").

Is there any way to retrieve the currently used client from a request/context?

**Comment by brockallen** [comment]
Out of curiosity, why would your API care how the client obtained the token? It is a bearer token, after all.

**Comment by StefanKoenigMUC** [comment]
We have to kinds of clients - users with roles and 'application' clients with full access rights with ClientCredentials.

As ClientCredentials don't have user roles (or at least, I cannot add scope 'role' to them - as I get an exception during startup) we had to differentiate there, as I was not able to figure out another/better solution for that. (still open for better aproaches)

**Comment by brockallen** [comment]
Either use client_id, the lack of a sub claim, or you can set custom claims on the Client config's ClientClaims.
